### ISSUE TITLE
GraphQL mapping for custom collections

### ISSUE THREAD

**Issue by Maxell92** [api-platform/core#2687]
We have several Doctrine entities linked like this: Order contains N boxes, each box contains several tracking numbers. I want to add a `getTrackingNumbers` method to Order Entity:

```
/**
 * @ApiResource
 *
 * @ORM\Entity
 */
class Order {
...
	/**
	 * @return TrackingNumber[]
	 */
	public function getTrackingNumbers()
	{
		$numbers = [];
		// some logic here

		return $numbers;
	}
```

Generated schema contains `trackingNumbers: Iterable!` instead of specific type. How can I fix annotation? It is working fine if I query boxes collection and then trackingNumbers collection, but I do not want to expose my DB schema and force the client to loop through properties he do not need. I tried to use `Subresource` annotation, but it did not helped.

**Comment by alanpoulain** [comment]
I'm working on a way to solve this kind of issue by adding a custom type to your schema and use it whenever you have a specific property in a specific class.

**Comment by Maxell92** [comment]
Thanks for answer. Not sure we need a Custom type - I am referencing known ApiResource here.

**Comment by alanpoulain** [comment]
I see. Maybe it's a real bug then. If you want to look it yourself the code handling it is here: https://github.com/api-platform/core/blob/be932e901402000adf54fa38e0b741f36474be79/src/GraphQl/Type/SchemaBuilder.php#L369-L386

**Comment by Maxell92** [comment]
Hi Alan,

thank you for directions. I tried to debug and this is what I have found:

* Property Extractor does not see method if there is no property with same name: https://github.com/api-platform/core/blob/be932e901402000adf54fa38e0b741f36474be79/src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php#L52
* So I added a property to my class:
```
	/**
	 * @var TrackingNumber[]|ArrayCollection
	 */
	private $trackingNumbers;
```
However Annotation is read by `PhpDocExtractor` and not by `DoctrineExtractor` in that case. https://github.com/api-platform/core/blob/be932e901402000adf54fa38e0b741f36474be79/src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php#L53 does not count with this different result:

**Doctrine extractor**
<img width="389" alt="Doctrine extractor" src="https://user-images.githubusercontent.com/1649279/55630943-32894780-57b7-11e9-96c6-2bba0c17cc8c.png">
	
**PhpDocExtractor**
<img width="427" alt="PhpDocExtractor" src="https://user-images.githubusercontent.com/1649279/55630965-3cab4600-57b7-11e9-8d9c-dfc4ae608c82.png">

I added code to merge information to `$type[0]` and that helped to generate correct schema. But once I want to query on this object, I hit a Doctrine error:
`[Syntax Error] line 0, col 61: Error: Unexpected 'WHERE'` in `QueryException.php:54`. I suppose that there is some code that tries to retrieve count and maybe performs JOIN?

Any hint how to solve this issue?

**Comment by Maxell92** [comment]
The code I used to test merging schema:
```
if (count($types) === 2 && (
	($types[0]->getBuiltinType() === Type::BUILTIN_TYPE_OBJECT && $types[1]->getBuiltinType() === Type::BUILTIN_TYPE_ARRAY) ||
	($types[1]->getBuiltinType() === Type::BUILTIN_TYPE_OBJECT && $types[0]->getBuiltinType() === Type::BUILTIN_TYPE_ARRAY)
	)
) {
	if ($types[0]->getBuiltinType() === Type::BUILTIN_TYPE_ARRAY) {
		$mainType = $types[0];
		$secondType = $types[1];
	}
	else {
		$mainType = $types[1];
		$secondType = $types[0];
	}

	$newType = new Type(
		Type::BUILTIN_TYPE_OBJECT,
		$mainType->isNullable(),
		$secondType->getClassName(),
		$mainType->isCollection(),
		$mainType->getCollectionKeyType(),
		$mainType->getCollectionValueType()
	);

	$types[0] = $newType;
}
```

**Comment by alanpoulain** [comment]
Hmm so it seems the behavior between the `DoctrineExtractor` and the `PhpDocExtractor` is really different for collections. Thanks for your investigation. It would help me if you could provide a failing Behat test in a PR :slightly_smiling_face:

**Comment by alanpoulain** [comment]
Should be solved with https://github.com/api-platform/core/pull/3327.

**Comment by nik-lampe** [comment]
@alanpoulain 

I have mentioned it on the slack channel, but it might be easer to track this here.

While creating an example for another problem I have with this annotation extracting thing I stumbled across this  Unexpected ‘WHERE error here.

Don’t know what I’m doing wrong here.

I’ve created an example:
https://github.com/nik-lampe/api-graphql-issue

I copypasted the book / review example from the docs. The only thing I have added is this:

https://github.com/nik-lampe/api-graphql-issue/blob/c305fa40cdd19c74c483b36f179213995e236736/src/Entity/Book.php#L80

```
    /**
     * @return Review[]
     */
    public function getBadReviews()
    {
        $badReviews = [];
        $reviews = $this->reviews;
        foreach ($reviews as $review) {
            if ($review->rating < 5) {
                $badReviews[] = $review;
            }
        }
        return $badReviews;
    }
```
